server/src/sim.rs

```diff
@@ -290,7 +389,12 @@ impl Sim {
         let _guard = span.enter();
 
         // Extend graph structure
-        for (_, (position, _)) in self.world.query::<(&mut Position, &mut Character)>().iter() {
+        for (_, (position, character)) in
+            self.world.query::<(&mut Position, &mut Character)>().iter()
+        {
+            if !character.state.active {
```

Comment:
I'm concerned about having to repeat this frequently, and risk bugs when it's forgotten. What if we instead introduced a new component `struct InactiveCharacter(Character);` and converted back and forth as needed?

Reply:
We don't currently have any machinery for adding and removing components in messages sent from the server to the client. However, the client doesn't have any reason to know the locations of inactive characters, so I could probably get away with just adding the entire entity to the "despawns" list when it gets deactivated. I'll at least try it, as I can always revert if necessary.

Reply:
In `snapshot_node`, I need to include the `Character` regardless of whether it's active or not. I don't think there's a good way of turning a `hecs::Ref<InactiveCharacter>` into a `hecs::Ref<Character>`, so it's not clear how I would want to implement `snapshot_node`. Calling another function for
```rust
components.push((ComponentType::Name as u64, ch.name.as_bytes().into()));
```
seems like overkill, but repeating all the code for the two component types seems too easy to get desynchronized.

There are risks with either implementation, since an inactive character might be ignored when it shouldn't be or treated as active when it shouldn't be. The simplicity of the boolean is enticing, and I'd probably need a bit more convincing to switch over to using another component.

Reply:
> I could probably get away with just adding the entire entity to the "despawns" list when it gets deactivated

That makes sense to me. No need for clients to know anything about inactive characters.

>  I don't think there's a good way of turning a hecs::Ref<InactiveCharacter> into a hecs::Ref<Character>

[`Ref::map`](https://docs.rs/hecs/latest/hecs/struct.Ref.html#method.map). With a little more finagling you could also operate in terms of `&T` instead of `Ref<T>`.

A more ECS-idiomatic solution would be to have a dedicated `Name` component that both disabled and active characters share.

> repeating all the code for the two component types seems too easy to get desynchronized.

Notwithstanding the above solutions, better to repeat `components.push(...)` once than `if !character.state.active { ... }` an unbounded number of times.

Reply:
That's a fair point. It's good to know that `Ref::map` exists. I might have to think more about this one, as I would like to try to get a self-consistent mental model that makes sense for this. It's a bit tricky because the server runs some kind of "simulation", but logging in and logging out causes players to temporarily disappear with the ability to reappear at the same spot later on with the same status.

In some games, like Satisfactory, logged out players still show up in the game world in their inactive state, in which case clients would want to know about them. That might be worth considering, in the spirit of allowing experimentation when we don't have a solid plan for gameplay.

I'm also second-guessing my decision to have `active` be a member of `Character`. It's possible we would want to replace a different component instead. For instance, having an `InactiveCharacter` component doesn't immediately resolve potential pitfalls because if we, for instance, added a `MinimapIcon` component (designed to mark things that should appear on the minimap) to the player, we would still need to remember to remove that component when players log out.

Another tricky thing is that, when saving, what we actually want to save is what the state of the simulation would be if everybody logged out, so the more differences there are between logged-in and logged-out players, the more potential room for error there is in the saving logic.

All in all, this is a tricky problem that I haven't really figured out what to do for. One part of me is thinking that logged-in and logged-out players should be treated as similarly as possible, leaving it up to the system logic to filter out inactive entities (possibly with an `Inactive` marker component), while the other part of me is thinking that logged-out players shouldn't be returned from ECS queries unless you are actively looking for them (which might involve hoisting them out of ECS entirely, although the question remains for what to do about the items they're holding).

If anything leans me one way, it would probably be towards the former due to the fact that we might want to do something with a logged-out player, such as having an admin move them or having the model rendered.

I'm definitely open to suggestions, though.

Reply:
I think removing any unwanted components on log-out is the idiomatic approach. That's a single code location to worry about and has highly predictable effects. Relying on every single system to understand and skip "inactive" entities has much more far reaching implications, and removing them entirely would make save/load more of a special case, especially for components (like `Inventory`) that might otherwise share logic with other entities.

Reply:
Which components are unwanted? I want `Position` because we need to remember where they were for when we log back in. I want `Character` because preserving the character state would allow things like velocity to be preserved if someone logs out. I want `Inventory` because players keep there items when logged out. The only component that is definitely unneeded is `CharacterInput`, since nobody is inputting anything when logged out.

However, if I just remove the `CharacterInput` component, other systems that don't care about that component would still need to understand and find some way to skip "inactive" entities.

One workaround you had suggested previously is an `InactiveCharacter`, but I'm not sure why it's okay to keep a real `Inventory` component when the `Character` component needs to be replaced.

That being said, I do think removing unneeded components is likely part of the solution. However, I still don't think there's a clean way to ensure that systems will never have to be aware of this "active" concept sometimes.

Since a lot of these decisions feel arbitrary, I think I might be more comfortable if I better understand the end goal. Why is the component called `Character` and not `Player`? Is the plan to reuse it for NPCs? Is the `name` field of the `Character` component only for players? When controlling an NPC's AI, if it interacts with other characters, what components will it care about?

I would like to have a good idea of some of those questions before I feel comfortable splitting the `Character` component. My understanding is that the idiomatic approach to ECS is for components to represent some kind of attribute the entity holds, and I think I'll need to know what those attributes are before I make too many changes. It doesn't help that after a player logs out, they're still conceptually a "character", just in some kind of stasis.

Honestly, requiring systems to be aware of the "active" state seems easier to refactor down the road (after we are actually seeing that repeated logic in enough places) than trying to design the components to make "active" and "inactive" come naturally. If you would prefer, I could still make the `InactiveCharacter` component, but I won't have a good intuition for it, so I wouldn't be able to design around it with any confidence.

Reply:
> Which components are unwanted?

Whichever ones are required by the logic you don't want to run.

> I want Character because preserving the character state would allow things like velocity to be preserved if someone logs out.

You want the information, not the compositional identity. Hence `struct InactiveCharacter(Character);`.

> other systems that don't care about that component would still need to understand and find some way to skip "inactive" entities.

If you remove the components they need, then you don't need to modify the other systems. What systems are you concerned about specifically?

> Why is the component called Character and not Player?

A player is a person. A character is an entity a player might control in a game world.

> Is the plan to reuse it for NPCs?

No specific plan here.

> Is the name field of the Character component only for players?

It seems like that might be convenient, so why not?

> I would like to have a good idea of some of those questions before I feel comfortable splitting the Character component.

There's only two fields; not a whole lot of splitting to agonize over. As a general rule of thumb, data should be gathered into a single component when it is always needed together, and shouldn't be when it isn't. When in doubt, keep things together for simplicity.

> requiring systems to be aware of the "active" state seems easier to refactor down the road

Most decisions at this stage are reasonably easy to refactor. They'll become less so as precedent sets in.

> after we are actually seeing that repeated logic in enough places

How many places is enough? We're already at four, counting the renderer.

Reply:
I think the best move here is for me to trust your judgement. If it's easy to refactor whatever the decision is, then I think I should stop agonizing over this and just make `InactiveCharacter`.

It might make sense to still transmit it to the client to make debugging easier by keeping the server and client as in-sync as possible. However, it might also be best to wait until a bug like this occurs that we want to debug. "Hidden information" in general seems like something we could want to support.

Reply:
Now that I've added the `InactiveCharacter` struct, I'll mark this as resolved.